design-proposals/klip-43-timestamp-data-type-support.md

```diff
@@ -0,0 +1,194 @@
+time data from other databases without having to convert column types, as well as simplify time data
+manipulation_
+
+## Motivation and background
+
+With Connect integration, KSQL processes data from various databases. Most other databases
+have a TIMESTAMP or DATETIME type, which KSQL currently converts to BIGINT. This makes sinking to
+databases that don't implicitly cast long values to timestamps much more complicated - the user would
+have to set up their connector to convert these columns back to TIMESTAMP. Supporting TIMESTAMP types
+would make moving time data between KSQL and other data sources/sinks smoother and less error-prone.
+
+Adding a TIMESTAMP type also simplifies time data manipulation. For example, currently
+if a user wants to extract the month of a timestamp, they would either have to parse a string or do a
+lot of math. Having time data in a dedicated data type allows for a lot of new UDFs.
+
+## What is in scope
+
+* Add TIMESTAMP type to KSQL
+* Support TIMESTAMP arithmetic and comparisons
+* Support TIMESTAMP usage in STRUCT, MAP and ARRAY
+* Serialization and de-serialization of TIMESTAMPs to Avro, JSON, Protobuf and Delimited formats
+* Update existing built-in functions to use the TIMESTAMP data type
+* Casting between TIMESTAMP and BIGINT
+
+## What is not in scope
+* Changing the ROWTIME data type. We will eventually want this to happen, but that is a separate
+discussion.
+* Support for dates before Unix epoch - this is not supported by Kafka. There is a [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-228+Negative+record+timestamp+support)
+for negative timestamps, but until that is implemented, KSQL can only support positive timestamps.
+* DATE and TIME types - because TIMESTAMPs represent a point in time, DATE and TIME types
+would be useful if a user wants to represent a less specific time. These would be added after
+TIMESTAMP is implemented though.
+* Allow window units (HOUR, DAY etc) to be used in timestamp functions.
+
+## Public APIS
+
+The TIMESTAMP data type will store a point in time after Unix epoch without timezone information.
+The syntax is as follows:
+
+```roomsql
+CREATE STREAM stream_name (time TIMESTAMP, COL2 STRING) AS ...
+CREATE TABLE table_name (col1 STRUCT<field TIMESTAMP>) AS ...
+```
+
+TIMESTAMPS will be displayed in console as strings in ISO-8601 format:
+
+```roomsql
+> SELECT time FROM stream_name EMIT CHANGES;
+
++------------------------+
+|time                    |
++------------------------+
+|1994-11-05T13:15:30Z    |
+```
+
+TIMESTAMPS can be represented by milliseconds from Unix epoch or by using conversion functions:
+
+```roomsql
+INSERT INTO stream_name VALUES (1605927509166);
+INSERT INTO stream_name VALUES (STRINGTOTIMESTAMP("2020-11-20", "YYYY-MM-DD"));
+```
+
+## Design
+
+### Serialization/Deserialization
+
+TIMESTAMPs will be handled by the `java.time.Instant` class within KSQL. The corresponding Kafka Connect type is
```

Comment:
I think it's important to have a discussion on time zone. Will all timestamps be treated by default in UTC? how could a user convert timestamps between zones?

Reply:
I was thinking that the timestamps here would reflect the timestamps in Kafka, which assumes UTC. MySQL and Snowflake have the `convert_tz` function, so we could do something similar.